### PR TITLE
Python mapper: Add region and geometry options

### PR DESCRIPTION
**Pull request by expertmm**
These changes make the script compatible with minetestmapper (binary)'s geometry param:
--geometry <xmin>:<zmin>+<width>+<height>
as well as adding a region param with the same values (but no spaces) as minetestmapper-numpy:
--region <xmin>:<xmax>,<zmin>:<zmax>

(other differences from master are not necessary)

### CODE REVIEW

File: util/minetestmapper.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
 #!/usr/bin/env python2
+# expertmm fork from 2016-02-01 git version
 # -*- coding: utf-8 -*-
 
 # This program is free software. It comes without any warranty, to
 # the extent permitted by applicable law. You can redistribute it
-# and/or modify it under the terms of the Do What The Fuck You Want
+# and/or modify it under the terms of the Do What ... You Want To
```
Comment (MoNTE48) [inline util/minetestmapper.py]: To-To :)

File: util/minetestmapper.py [hunk 2]
```diff
@@ -1,9 +1,10 @@
 #!/usr/bin/env python2
+# expertmm fork from 2016-02-01 git version
```
Comment (est31) [inline util/minetestmapper.py]: You should put the attribution to you to where the other people are attributed. There is a list of changes, add yours to it.

File: util/minetestmapper.py [hunk 3]
```diff
@@ -151,19 +154,16 @@ def usage():
 path = None
 output = "map.png"
 border = 0
-scalecolor = "black"
-bgcolor = "white"
+scalecolor = "white"
```
Comment (est31) [inline util/minetestmapper.py]: why did you modify this? This is unrelated to the change.

File: util/minetestmapper.py [hunk 4]
```diff
@@ -190,8 +190,77 @@ def usage():
...
+            print("#  height:"+str(this_height))
+            print("region:")
+            print("  xmin:"+str(nonchunky_xmin))
+            print("  xmax:"+str(nonchunky_xmax))
+            print("  zmin:"+str(nonchunky_zmin))
+            print("  zmax:"+str(nonchunky_zmax))
+        else:
+            print("ERROR: (Missing coordinates in '"+geometry_string+"' for geometry) Geometry should be in the form: x:z+width+height")
+            usage()
+            sys.exit()
```
Comment (est31) [inline util/minetestmapper.py]: also put a non zero exit status, it is important for script usage

File: util/minetestmapper.py [hunk 5]
```diff
@@ -190,8 +190,77 @@ def usage():
...
+    else:
+        print("ERROR: (Incorrect value '"+geometry_string+"' for geometry) Geometry should be in the form: x:z+width+height")
+        usage()
+        sys.exit()
```
Comment (est31) [inline util/minetestmapper.py]: same here

**Comment by expertmm** [comment]
I made those changes, thank you.

File: util/minetestmapper.py (hunk 2 above)
Comment (expertmm) [inline util/minetestmapper.py]: ok done thank you for pointing that out

File: util/minetestmapper.py (hunk 3 above)
Comment (expertmm) [inline util/minetestmapper.py]: Ok, I reverted those changes. Those were related to my usage, but now I set color in my program instead.

File: util/minetestmapper.py (hunk 4 above)
Comment (expertmm) [inline util/minetestmapper.py]: done.

File: util/minetestmapper.py (hunk 5 above)
Comment (expertmm) [inline util/minetestmapper.py]: done.

File: util/minetestmapper.py (hunk 1 above)
Comment (expertmm) [inline util/minetestmapper.py]: ok fixed now ty

File: util/minetestmapper.py [hunk 6]
```diff
@@ -190,8 +190,77 @@ def usage():
...
+            print("#  z:"+str(z))
+            print("#  width:"+str(this_width))
+            print("#  height:"+str(this_height))
+            print("region:")
+            print("  xmin:"+str(nonchunky_xmin))
+            print("  xmax:"+str(nonchunky_xmax))
+            print("  zmin:"+str(nonchunky_zmin))
+            print("  zmax:"+str(nonchunky_zmax))
+        else:
+            print("ERROR: (Missing coordinates in '"+geometry_string+"' for geometry) Geometry should be in the form: x:z+width+height")
```
Comment (kwolekr) [inline util/minetestmapper.py]: This line looks a little too long, for PEP8 compliance it needs to be <= 80 columns.

File: util/minetestmapper.py [hunk 7]
```diff
@@ -190,8 +190,77 @@ def usage():
...
+            x = int(x_string)
+            z = int(z_string)
+            this_width = int(width_string)
+            this_height = int(height_string)
+            nonchunky_xmin = x
+            nonchunky_xmax = nonchunky_xmin + this_width - 1  # -1 since max is inclusive rect
+            nonchunky_zmin = z
+            nonchunky_zmax = nonchunky_zmin + this_height - 1  # -1 since max is inclusive rect
+            print("#geometry:")
+            print("#  x:"+str(x))
```
Comment (kwolekr) [inline util/minetestmapper.py]: Spaces between operators?

File: util/minetestmapper.py [hunk 8]
```diff
@@ -190,8 +190,77 @@ def usage():
...
+
+if geometry_string is not None:
+    parts = geometry_string.split("+")
+    if len(parts) == 3:
+        coords_string, width_string, height_string = parts
+        coords = coords_string.split(":")
+        if len(coords)==2:
+            x_string, z_string = coords
+            x = int(x_string)
+            z = int(z_string)
```
Comment (kwolekr) [inline util/minetestmapper.py]: What about a list comprehension or lambda here?
`x, z = [int(foo) for foo in coords]`

File: util/minetestmapper.py [hunk 9]
```diff
@@ -202,8 +271,30 @@ def usage():
...
 colors = {}
+colors_path = "colors.txt"
+
+os_name="linux"
+if (os.path.sep!="/"):
+    os_name="windows"
+
+profile_path = None
+if os_name=="windows":
+    profile_path = os.environ['USERPROFILE']
```
Comment (kwolekr) [inline util/minetestmapper.py]: ?? I'm pretty sure this isn't necessary, Windows has %HOME% as an alias for %USERPROFILE%

File: util/minetestmapper.py [hunk 10]
```diff
@@ -202,8 +271,30 @@ def usage():
...
+else:
+    profile_path = os.environ['HOME']
+
+mt_path = os.path.join( profile_path, "minetest")
```
Comment (kwolekr) [inline util/minetestmapper.py]: Isn't the minetest directory ~/.minetest?

File: util/minetestmapper.py [hunk 11]
```diff
@@ -131,6 +132,8 @@ def readS32(f):
   --drawscale
   --drawplayers
   --draworigin
+  --region <xmin>:<xmax>,<zmin>:<zmax>
+  --geometry <xmin>:<zmin>+<width>+<height>
   --drawunderground
 Color format: '#000000'"""
```
Comment (kwolekr) [inline util/minetestmapper.py]: Would be nice to have something here talk about the colors.txt file and where this script tries to find it.

File: util/minetestmapper.py [hunk 12]
```diff
@@ -202,8 +271,30 @@ def usage():
 
 # Load color information for the blocks.
 colors = {}
+colors_path = "colors.txt"
+
+os_name="linux"
+if (os.path.sep!="/"):
```
Comment (ShadowNinja) [inline util/minetestmapper.py]: I'm sure there's a better vay to do this test.  `os` uses `"nt" in sys.builtin_module_names`.
Also `s/linux/posix/`.

File: util/minetestmapper.py [hunk 13]
```diff
@@ -202,8 +271,30 @@ def usage():
...
+
+mt_path = os.path.join( profile_path, "minetest")
+mt_util_path = os.path.join( mt_path, "util")
+abs_colors_path = os.path.join( mt_util_path, "colors.txt" )
+
+if not os.path.isfile(colors_path):
+    colors_path = abs_colors_path
+
+if not os.path.isfile(colors_path):
+    print("WARNING: could not find colors.txt is current path or '"+colors_path+"'")
```
Comment (ShadowNinja) [inline util/minetestmapper.py]: `s/is/in/`

File: util/minetestmapper.py [hunk 14]
```diff
@@ -284,7 +375,7 @@ def usage():
         zlist.append(z)
 
 if len(xlist) == 0 or len(zlist) == 0:
-    print("World does not exist.")
+    print("At this location, data does not exist.")
```
Comment (ShadowNinja) [inline util/minetestmapper.py]: `"Data does not exist at this location."`

**Comment by ShadowNinja** [comment]
The `--geometry` format really isn't all that good.  `:,:` or just `:::`/`,,,` would be better.

**Comment by kwolekr** [comment]
@ShadowNinja that geometry format is standard throughout *nix, I think it should remain the way it is now.

**Comment by ShadowNinja** [comment]
@kwolekr: Really?  Imagemagik has a `<width>x<height>{+-}<xoffset>{+-}<yoffset>` system, but that's all I saw with a quick web search.

**Comment by expertmm** [comment]
I made changes in response to previous comments (see title and description of commits above)

**Comment by expertmm** [comment]
It is still marked "Action needed," but I think all work that was suggested is done. Anyone have comments on this version?

File: util/minetestmapper.py [hunk 15]
```diff
@@ -190,22 +199,126 @@ def usage():
...
+                  "' for geometry (must be in the form: x:z+width+height)")
+            usage()
+            sys.exit(2)
+    else:
+        print("ERROR: Incorrect value '" + geometry_string +
+              "' for geometry (must be in the form: x:z+width+height)")
+        usage()
+        sys.exit(2)
+elif region_string is not None:
+    # parts = region_string.split(" ")
```
Comment (kwolekr) [inline util/minetestmapper.py]: ??

File: util/minetestmapper.py [hunk 16]
```diff
@@ -190,22 +199,126 @@ def usage():
...
+mt_util_path = None
+abs_colors_path = None
+
+if not os.path.isfile(colors_path):
+    colors_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                               "colors.txt")
+
+if not os.path.isfile(colors_path):
+    if profile_path is not None:
+        try_path = os.path.join(profile_path, "minetest")
```
Comment (kwolekr) [inline util/minetestmapper.py]: Are you sure this is the correct directory name?

**Comment by kwolekr** [comment]
Looking good expertmm.  Just want to make sure on the profile dir.

ShadowNinja, Xorg uses the format extensively.

File: util/minetestmapper.py (hunk 16 above)
Comment (expertmm) [inline util/minetestmapper.py]: I thought using the profile_path would be ok since it is now just for fault tolerance (for one step in finding try_path), while a colors.txt in the same folder takes priority, in which case profile_path will not be used. Let me know if you still think it should be changed.

**Comment by expertmm** [comment]
I'm not sure what other action may be needed--I'm pretty sure I did everything required. Let me know if I should remove that profile code or not, or whether my changes are satisfactory. I removed the license completely this time since it seems that the file should really fall under the minetest license anyway, like the other py files, and since recently developers brought up how WTFPL is not a legally safe license (it was discussed that MIT and CC-BY-3.0 and even CC0 offer more adequate protections).

File: util/minetestmapper.py [hunk 17]
```diff
@@ -190,22 +194,126 @@ def usage():
...
+            x_string, z_string = coords
+            x = int(x_string)
+            z = int(z_string)
+            this_width = int(width_string)
+            this_height = int(height_string)
+            nonchunky_xmin = x
+            nonchunky_xmax = nonchunky_xmin + this_width - 1  # inclusive rect
+            nonchunky_zmin = z
+            nonchunky_zmax = nonchunky_zmin + this_height - 1  # inclusive rect
+            print("#geometry:")
```
Comment (nerzhul) [inline util/minetestmapper.py]: call print once using multiline comment

Comment (expertmm) [inline util/minetestmapper.py]: If you are saying something about the '#'s being printed, that is so the output is machine-readable (always prints region notation, and prints comments with values from geometry notation if that was the source). I can't call print once using a multiline comment (string) since there are variables. I'm not sure what you mean. Do you mean formatted multiline string like:
```python
print("""
#geometry:
#    x:{}
#    z:{}
#    width:{}
#    height:{}
region:
  xmin:{}
  xmax:{}
  zmin:{}
  zmax:{}
                  """.format(x,z,this_width,this_height,
                             nonchunky_xmin,nonchunky_xmax,
                             nonchunky_zmin,nonchunky_zmax))
```
(leaving indents off of the string as done above would be necessary due to the fact they would be printed otherwise--the way around this would be adding the lines together, keeping the string conversions [or format call for each line] and putting a newline at the end of each instead of using a multiline string)

Comment (expertmm) [inline util/minetestmapper.py]: (or
```python

            print("#geometry:"+"\n"
                +"#  x:" + str(x)+"\n"
                +"#  z:" + str(z)+"\n"
                +"#  width:" + str(this_width)+"\n"
                +"#  height:" + str(this_height)+"\n"
                +"region:"+"\n"
                +"  xmin:" + str(nonchunky_xmin)+"\n"
                +"  xmax:" + str(nonchunky_xmax)+"\n"
                +"  zmin:" + str(nonchunky_zmin)+"\n"
                +"  zmax:" + str(nonchunky_zmax))
```
)

Comment (expertmm) [inline util/minetestmapper.py]: or
```
            print("#geometry:\n" + \
                  "#  x:{}\n".format(x) + \
                  "#  z:{}\n".format(z) + \
                  "#  width:{}\n".format(this_width) + \
                  "#  height:{}\n".format(this_height) + \
                  "region:\n" + \
                  "  xmin:{}\n".format(nonchunky_xmin) + \
                  "  xmax:{}\n".format(nonchunky_xmax) + \
                  "  zmin:{}\n".format(nonchunky_zmin) + \
                  "  zmax:{}\n".format(nonchunky_zmax))
```
or
```python
            print("#geometry:\n"
                  "#  x:{}\n"
                  "#  z:{}\n"
                  "#  width:{}\n"
                  "#  height:{}\n"
                  "region:\n"
                  "  xmin:{}\n"
                  "  xmax:{}\n"
                  "  zmin:{}\n"
                  "  zmax:{}\n".format(x,z,this_width,this_height,
                                       nonchunky_xmin,nonchunky_xmax,
                                       nonchunky_zmin,nonchunky_zmax))
```
there are so many ways, but that last one seems the most readable and keeps indentation neat.

**Comment by expertmm** [comment]
Sorry for the multiple posts. I just was not sure what you wanted and there is much code in there that doesn't conform to PEP8. After conforming my code to PEP8, I noticed much more was also not conformed so I just used [http://pep8online.com](http://pep8online.com) and conformed all of the original code. Let me know if there are any other issues.

**Comment by expertmm** [comment]
Can someone check whether this is adequate? I believe I made all changes requested, and made the other code conform to pep8 as well.

**Comment by ShadowNinja** [comment]
Python mapper removed.
